File: src/main/java/org/distril/beengine/material/block/Block.java
package org.distril.beengine.material.block;

import com.nukkitx.math.vector.Vector3i;
import lombok.Getter;
import lombok.Setter;
import org.distril.beengine.material.Material;
import org.distril.beengine.util.Direction;
import org.distril.beengine.world.World;

import java.util.Objects;

@Getter
@Setter
public abstract class Block implements Cloneable, BlockBehaviors {

	private final Material material;

	private BlockState state;

	private World world;
	private Vector3i position;

	public Block(Material material) {
		this.material = material;
		this.state = BlockPalette.getDefaultState(material);
	}

	public Block(Material material, BlockState state) {
		this.material = material;
		this.state = state;
	}

	public void setPosition(World world, Vector3i position) {
		this.world = world;
		this.position = position;
	}

	public Block getSide(Direction face) {
		return this.getSide(face, 1);
	}

	public Block getSide(Direction face, int step) {
		return this.world.getBlock(face.getOffset(this.position, step));
	}

	@Override
	public Block clone() {
		try {
			Block clone = (Block) super.clone();
			clone.state = state;
			clone.world = world;
			clone.position = position;
			return clone;
		} catch (CloneNotSupportedException exception) {
			throw new AssertionError();
		}
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj) {
			return true;
		}

		if (obj == null || this.getClass() != obj.getClass()) {
			return false;
		}

		Block that = (Block) obj;
		return this.material == that.getMaterial() && this.state.equals(that.getState());
	}

	@Override
	public int hashCode() {
		return Objects.hash(this.material, this.state);
	}
}

File: src/main/kotlin/org/distril/beengine/material/block/Block.kt
package org.distril.beengine.material.block

import com.nukkitx.math.vector.Vector3i
import org.distril.beengine.material.Material
import org.distril.beengine.util.Direction
import org.distril.beengine.world.World
import java.util.*

abstract class Block(val material: Material, state: BlockState?) : Cloneable, BlockBehaviors {

	var state = state ?: BlockPalette.getDefaultState(material)!!

	var world: World? = null
	var position: Vector3i? = null

	fun getSide(face: Direction, step: Int = 1) = this.world!!.getBlock(face.getOffset(this.position!!, step))

	public override fun clone(): Block {
		val clone = super.clone() as Block
		clone.state = this.state
		clone.world = this.world
		clone.position = this.position
		return clone
	}

	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (other !is Block) return false

		return this.material === other.material && this.state == other.state
	}

	override fun hashCode() = Objects.hash(this.material, this.state)
}
